tests/unit/resources/test_compensation.py

```diff
@@ -98,3 +108,72 @@ def test_create_from_spill_string(spillstring):
         "Qdot655-A",
         "Qdot705-A",
     ]
+
+
+@responses.activate
+def test_apply_comp_errors_for_nonmatching_channels(
+    client, ENDPOINT_BASE, compensation, fcs_file
+):
+    events_body = open("tests/data/Acea - Novocyte.fcs", "rb")
+    responses.add(
+        responses.GET,
+        f"{ENDPOINT_BASE}/experiments/{EXP_ID}/fcsfiles/{fcs_file._id}.fcs",
+        body=events_body,
+    )
+    events = parse_fcs_file(client.download_fcs_file(EXP_ID, fcs_file._id))
+    fcs_file.events = events
+
+    with pytest.raises(IndexError):
+        compensation.apply(fcs_file)
+
+
+@responses.activate
+def test_apply_compensation_to_fcs_file_with_matching_kwargs(
+    client, ENDPOINT_BASE, compensation, fcs_file
+):
+    # Given: a Compensation with channels as a subset of the FcsFile events
+    responses.add(
+        responses.GET,
+        f"{ENDPOINT_BASE}/experiments/{EXP_ID}/fcsfiles/{fcs_file._id}.fcs",
+        body=open("tests/data/Acea - Novocyte.fcs", "rb"),
+    )
+    events = fcs_file.get_events(inplace=True, testKwarg="foo")
+    assert fcs_file._events_kwargs == {"testKwarg": "foo"}
+
+    ix = list(events.columns)
+    compensation.dataframe = DataFrame(identity(24), index=ix, columns=ix)
+    compensation.channels = ix
+
+    # When: a Compensation is applied
+    results = compensation.apply(fcs_file, testKwarg="foo")
+
+    # Then: events should be compensated
+    assert all(results == events)
+    assert (
+        responses.assert_call_count(
+            f"{ENDPOINT_BASE}/experiments/{EXP_ID}/fcsfiles/{fcs_file._id}.fcs?testKwarg=foo",
+            1,
+        )
+        is True
+    )
+
+
+@responses.activate
+def test_apply_comp_compensates_values(
+    acea_events_compensated,
+    acea_fcs_file,
+    acea_compensation,
+):
+    """This test compares results from a file-internal compensation conducted
+    by the Python toolkit to one conducted by CellEngine. See
+    tests/fixtures/compensated_events.py for details on the fixtures used
+    here."""
+    # Given:
+    # - a file-internal compensation (see tests/fixtures/compensated_events.py)
+    # - an FcsFile with uncompensated events
+
+    # When: the Compensation is applied to a file
+    results = acea_compensation.apply(acea_fcs_file, inplace=False)
+
+    # Then: events should be compensated correctly
+    assert all(results.head(5) == acea_events_compensated.head(5))
```

Comment:
This test doesn't seem to be asserting correctly. It looked suspicious that these were digit-for-digit identical between CellEngine and Python. I changed a few numbers in acea_compensated.json and the test still passes.
![image](https://user-images.githubusercontent.com/469365/144933566-0ccc7faf-0e4f-4935-b479-0500879e64e8.png)

Reply:
Possibly a good use case for [`pandas.testing.assert_frame_equal`](https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html) , which supports approximate (with tolerances) matching.